tree/treeplayer/src/TTreeProcessorMT.cxx

```diff
@@ -195,6 +195,52 @@ static std::vector<std::string> GetTreeFullPaths(const TTree &tree)
    return {tree.GetName()};
 }
 
+/// Convert a range start/end of global entry numbers to the corresponding entry numbers in the TEntryList elist.
+///
+/// This routine assumes that entry numbers in the TEntryList (and, if present, in the sub-entrylists) are in
+/// ascending order, i.e., for n > m:
+///   elist.GetEntry(n) + tree_offset_for_entry_from_elist(n) > elist.GetEntry(m) + tree_offset_for_entry_from_elist(m)
+///
+/// For TChains and TEntryLists with sub-entrylists, this routine requires retrieval of the number of entries in each
+/// tree in the chain in order to covert local entry numbers into global entry numbers.
+static std::pair<Long64_t, Long64_t> ToEntryListRange(Long64_t start, Long64_t end, TEntryList &elist, TChain &chain)
+{
+   Long64_t elistStart = -1ll, elistEnd = -1ll;
+
+   const auto nEntries = elist.GetN();
+   for (auto e = 0ll; e < nEntries; ++e) {
```

Comment:
We could leave this as is *but* this looks very inefficient for large non-sparse list.
Moving this inside TEntryList could be much faster (See TEntryList::Contains for hints why).   This will require tweaks to TEntryListBlock (it has a fast interface to find out if an entry is in the list but not (yet) what is its "index" in the list. [Humm reading TEntryList::GetEntry, the translation from "block info" to "index" is ... well .. indirect ... so maybe a more indirect route (looping over list and blocks rather elements)]

Reply:
You are right, it is quite inefficient...what I can do, now that you make me think about it, is to perform this entry number translation once for all clusters rather than once per task/cluster. I'll fix this.